### Пример 1

### Было

```kotlin
//ViewModel (UI-слой)
fun getProfileData() {
    viewModelScope.launch {
        try {
            val profileData = getProfileDataUseCase()
            _profileDataState.emit(ProfileDataUI.fromProfileData(profileData))
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }
}

//data-слой
class ProfileDataRepositoryImpl @Inject constructor(
    private val profileDataApi: ProfileDataApi
): ProfileDataRepository {
    override suspend fun getProfileData(): ProfileData {
        return profileDataApi.getProfileData().first().toModel()
    }

}

```

### Стало

```kotlin
//data-слой
class ProfileDataRepositoryImpl @Inject constructor(
    private val profileDataApi: ProfileDataApi,
    private val defaultDispatcher: CoroutineDispatcher,
): ProfileDataRepository {
    
    override suspend fun getProfileData(): Flow<Result<ProfileData>> = flow {
        emit(Result.Loading)
        val result = profileDataApi.getProfileData().first().toModel()
        emit(Result.Success(result))
    }.catch {
        emit(Result.Failure(it.fillInStackTrace()))
    }.flowOn(defaultDispatcher)

}

sealed class Result<out T> {
    object Loading : Result<Nothing>()
    class Failure(val msg: Throwable) : Result<Nothing>()
    class Success<T>(val data: T) : Result<T>()
    object Empty : Result<Nothing>()
}

fun getProfileData() {
    getProfileDataUseCase().onEach {
        when (it) {
            is Result.Success -> emitProfileData(it.data)
            is Result.Failure -> printError(it.msg.localizedMessage)
            is Result.Loading -> processLoading()
            is Result.Empty -> processEmptyResult()
        }
    }.launchIn(viewModelScope)
}

fun emitProfileData(profileData: ProfileData){}

fun printError(message: String?){}

fun processLoading(){}

fun processEmptyResult(){}

```

Во множестве мест проекта для отлавливания ошибок использовалась конструкция try-catch
в слое ui, при этом она отлавливала любые исключения "на всякий случай", и никак их не
обрабатывала.

В исправленной версии кода добавил строго определенное множество ситуаций, каждую из
которых нужно обработать, иначе будет ошибка компиляции. Таким образом возросла
предсказуемость кода и он стал больше соответствовать логическому дизайну системы.